File: backend/routes/azienda.js
const express = require('express');
const router = express.Router();
const db = require('../database');
const { requireRole } = require('../middleware/auth');

const ADMIN_ROLES = ['SUPERADMIN', 'ADMIN'];

router.get('/', (req, res) => {
  const row = db.prepare('SELECT * FROM azienda WHERE id = 1').get();
  if (!row) return res.json({});
  const isAdmin = req.user && ADMIN_ROLES.includes(req.user.ruolo);
  res.json(toDto(row, isAdmin));
});

router.put('/', requireRole('SUPERADMIN', 'OWNER', 'ADMIN'), (req, res) => {
  const a = req.body;
  db.prepare(`UPDATE azienda SET ragione_sociale=?, indirizzo=?, cap=?, citta=?, provincia=?, stato=?,
    p_iva=?, cod_fiscale=?, email=?, telefono=?, pec=?, sdi=?, banca=?, iban=?, logo=?,
    smtp_host=?, smtp_port=?, smtp_user=?, smtp_pass=?, smtp_from=?, smtp_secure=?,
    sdi_api_url=?, sdi_api_key=?,
    riordino_automatico=?, multi_utente_attivo=?,
    numerazione_annuale=?, numero_prefissi=?,
    template_config=?, notifiche_config=?, email_corpo_documento=?, email_mode=?,
    lock_documenti_default=?
    WHERE id=1`)
    .run(a.ragioneSociale, a.indirizzo, a.cap, a.citta, a.provincia, a.stato,
         a.pIva, a.codFiscale, a.email, a.telefono, a.pec, a.sdi, a.banca, a.iban, a.logo || '',
         a.smtpHost || '', a.smtpPort || 587, a.smtpUser || '', a.smtpPass || '', a.smtpFrom || '', a.smtpSecure ? 1 : 0,
         a.sdiApiUrl || '', a.sdiApiKey || '',
         a.riordinoAutomatico ? 1 : 0, a.multiUtenteAttivo ? 1 : 0,
         a.numerazioneAnnuale ? 1 : 0, JSON.stringify(a.numeroPrefissi || {}),
         a.templateConfig ? JSON.stringify(a.templateConfig) : null,
         a.notificheConfig ? JSON.stringify(a.notificheConfig) : null,
         a.emailCorpoDocumento ?? null,
         ['SMTP','MAILTO','WEBMAIL_GMAIL','WEBMAIL_OUTLOOK'].includes(a.emailMode) ? a.emailMode : 'SMTP',
         a.lockDocumentiDefault === false ? 0 : 1);
  res.json({ success: true });
});

function toDto(r, includeSecrets = false) {
  const maskedPass = r.smtp_pass ? '••••••••' : '';
  const maskedKey  = r.sdi_api_key ? '••••••••' : '';
  return {
    id: r.id, ragioneSociale: r.ragione_sociale, indirizzo: r.indirizzo,
    cap: r.cap, citta: r.citta, provincia: r.provincia, stato: r.stato,
    pIva: r.p_iva, codFiscale: r.cod_fiscale,
    email: r.email, telefono: r.telefono, pec: r.pec, sdi: r.sdi,
    banca: r.banca, iban: r.iban, logo: r.logo || '',
    smtpHost: r.smtp_host || '', smtpPort: r.smtp_port || 587,
    smtpUser: r.smtp_user || '', smtpPass: includeSecrets ? (r.smtp_pass || '') : maskedPass,
    smtpFrom: r.smtp_from || '', smtpSecure: r.smtp_secure === 1,
    sdiApiUrl: r.sdi_api_url || '', sdiApiKey: includeSecrets ? (r.sdi_api_key || '') : maskedKey,
    riordinoAutomatico: r.riordino_automatico === 1,
    multiUtenteAttivo: r.multi_utente_attivo === 1,
    numerazioneAnnuale: (r.numerazione_annuale ?? 1) !== 0,
    numeroPrefissi: (() => { try { return JSON.parse(r.numero_prefissi || '{}'); } catch(_) { return {}; } })(),
    templateConfig: (() => { try { return r.template_config ? JSON.parse(r.template_config) : null; } catch(_) { return null; } })(),
    notificheConfig: (() => { try { return r.notifiche_config ? JSON.parse(r.notifiche_config) : null; } catch(_) { return null; } })(),
    emailCorpoDocumento: r.email_corpo_documento || '',
    emailMode: ['SMTP','MAILTO','WEBMAIL_GMAIL','WEBMAIL_OUTLOOK'].includes(r.email_mode) ? r.email_mode : 'SMTP',
    lockDocumentiDefault: (r.lock_documenti_default ?? 1) !== 0,
  };
}

module.exports = router;

File: OrdevaDesktop/Models/Azienda.cs
namespace Ordeva.Desktop.Models;

/// <summary>
/// Impostazioni azienda: riga singleton (id=1) della tabella <c>azienda</c>.
/// Specchio esatto delle colonne SQLite (snake_case) in PascalCase.
/// I segreti (SmtpPass, SdiApiKey) vengono restituiti mascherati dalla repo,
/// esattamente come faceva il backend per gli utenti non SUPERADMIN/ADMIN.
/// </summary>
public sealed class Azienda
{
    /// <summary>Placeholder mostrato al posto dei segreti già impostati.</summary>
    public const string MaskedSecret = "••••••••";

    public long Id { get; set; } = 1;

    // ── Dati societari ───────────────────────────────────────────────
    public string RagioneSociale { get; set; } = "";
    public string PIva { get; set; } = "";
    public string CodFiscale { get; set; } = "";

    // ── Regime fiscale / ritenute / cassa ────────────────────────────
    public string RegimeFiscale { get; set; } = "RF01";
    public decimal RitenutaAliquotaDefault { get; set; }
    public string RitenutaTipoDefault { get; set; } = "RT02";
    public string RitenutaCausaleDefault { get; set; } = "";
    public string CassaTipoDefault { get; set; } = "";
    public decimal CassaAliquotaDefault { get; set; }
    public decimal CassaIvaDefault { get; set; }

    // ── Sede legale ──────────────────────────────────────────────────
    public string Indirizzo { get; set; } = "";
    public string Cap { get; set; } = "";
    public string Citta { get; set; } = "";
    public string Provincia { get; set; } = "";
    public string Stato { get; set; } = "";

    // ── Contatti ─────────────────────────────────────────────────────
    public string Telefono { get; set; } = "";
    public string Email { get; set; } = "";
    public string Pec { get; set; } = "";
    public string Sdi { get; set; } = "";

    // ── Dati bancari ─────────────────────────────────────────────────
    public string Iban { get; set; } = "";
    public string Banca { get; set; } = "";

    // ── Logo (base64/data-url) ───────────────────────────────────────
    public string Logo { get; set; } = "";

    // ── Email / SMTP ─────────────────────────────────────────────────
    public string EmailMode { get; set; } = "SMTP";
    public string SmtpHost { get; set; } = "";
    public int SmtpPort { get; set; } = 587;
    public string SmtpUser { get; set; } = "";
    public string SmtpPass { get; set; } = "";
    public string SmtpFrom { get; set; } = "";
    public bool SmtpSecure { get; set; }
    public string EmailCorpoDocumento { get; set; } = "";

    // ── SDI / fatturazione elettronica ───────────────────────────────
    public string SdiProvider { get; set; } = "GENERICO";
    public string SdiApiUrl { get; set; } = "";
    public string SdiApiKey { get; set; } = "";

    // ── Comportamenti ────────────────────────────────────────────────
    public bool RiordinoAutomatico { get; set; }
    public bool MultiUtenteAttivo { get; set; }
    public bool NumerazioneAnnuale { get; set; } = true;
    public bool LockDocumentiDefault { get; set; } = true;

    // ── Blob JSON: conservati grezzi (la UI desktop non li edita ancora) ──
    /// <summary>JSON dei prefissi numerazione documenti. Default "{}".</summary>
    public string NumeroPrefissi { get; set; } = "{}";
    /// <summary>JSON config grafica template stampa, o null.</summary>
    public string? TemplateConfig { get; set; }
    /// <summary>JSON config notifiche, o null.</summary>
    public string? NotificheConfig { get; set; }
}
